abdiff/core/collate_ab_transforms.py

```diff
@@ -166,90 +182,99 @@ def get_transformed_batches_iter(
-           is empty.
+    As records are joined, an "abdiff_record_id" UUID is minted to unambiguously reference
+    that particular row, which is helpful later during deduping.
 
-    The returned generator can be passed to abdiff.core.utils.write_to_dataset()
-    to perform batch writes of joined TIMDEX A/B records to a Parquet dataset
+    As the records are joined, record batches are yielded which get written to a new,
+    temporary "joined" dataset on disk.
 
     Args:
         dataset_directory: The root directory of the Parquet dataset of TIMDEX records
             (i.e., the tempfile.TemporaryDirectory).
     """
+    partition_paths = glob.glob(f"{dataset_directory}/transformed_file_name=*")
+    transformed_file_names = [os.path.basename(p).split("=")[1] for p in partition_paths]
+
     with duckdb.connect(":memory:") as con:
-        transformed_file_names = con.execute(
-            """
-            SELECT DISTINCT transformed_file_name FROM
-            read_parquet($transformed_parquet_glob, hive_partitioning=true)
-            """,
-            {"transformed_parquet_glob": f"{dataset_directory}/**/*.parquet"},
-        ).fetch_df()["transformed_file_name"]
+        for i, transformed_file in enumerate(transformed_file_names):
+            logger.debug(
+                f"Joining records batch {i+1}/{len(transformed_file_names)}: "
+                f"{transformed_file}"
+            )
+
+            transformed_file_parquet_glob = (
+                f"{dataset_directory}/transformed_file_name={transformed_file}/*.parquet"
+            )
 
-        for transformed_file in transformed_file_names:
             results = con.execute(
                 """
                 WITH
                     transformed_file AS (
                         SELECT * FROM
                         read_parquet(
-                            $transformed_parquet_glob,
+                            $transformed_file_parquet_glob,
                             hive_partitioning=true
                         )
-                        WHERE transformed_file_name=$transformed_file_name
                     ),
                     a AS (SELECT * FROM transformed_file WHERE version='a'),
                     b AS (SELECT * FROM transformed_file WHERE version='b')
                 SELECT
-                    COALESCE(a.timdex_record_id, b.timdex_record_id) timdex_record_id,
-                    COALESCE(a.source, b.source) source,
-                    COALESCE(a.run_date, b.run_date) run_date,
-                    COALESCE(a.run_type, b.run_type) run_type,
-                    COALESCE(a.action, b.action) "action",
-                    a.record as record_a,
-                    b.record as record_b
+                    uuid() as abdiff_record_id,
+                    a.timdex_record_id,
+                    a.source,
+                    a.run_date,
+                    a.run_type,
```

Comment:
For all these values -- `timdex_record_id`, `source`, `run_date`, and `run_type` -- we are gauranteed they will be the same value per this join.  As such, we can remove othe unecessary `COALESCE()` and just take the value from either `a` or `b`.  I missed this on our first pass on this.  

It's probably inconsequential, but at scale removing 4 `COALESCE()` calls should be a bit quicker.

Reply:
@jonavellecuerdo - even if a record is "deleted" in version B, it would still have column values for `timdex_record_id`, `source`, `run_date`, and `run_type`.  At least it should: these can all be known from the filename of the "<source>-<date>-<run-type>-to-delete.txt" that Transmogrifier produces.  

If not, then you're right; it'd be dangerous to arbitrarily only take these columns from A or B.

And I think we have proof that these will always be set [from here](https://github.com/MITLibraries/transmogrifier-ab-diff/blob/TIMX-383-pipeline-tweaks-large-runs/abdiff/core/collate_ab_transforms.py#L138-L140) where we produce the base of the record, regardless of "to-index" or "to-delete", ensuring these values are always present.  These values come from the `filename_details`, which is produced by `parse_timdex_filename()`.